pkg/operator/ceph/cluster/osd/provision_spec.go

```diff
@@ -157,17 +157,31 @@ func (c *Cluster) provisionPodTemplateSpec(osdProps osdProperties, restart v1.Re
 		podSpec.DNSPolicy = v1.DNSClusterFirstWithHostNet
 	}
 	if osdProps.onPVC() {
-		// The "all" placement is applied separately so it will have lower priority.
-		// We want placement from the storageClassDeviceSet to be applied and override
-		// the "all" placement if there are any overlapping placement settings.
-		c.spec.Placement.All().ApplyToPodSpec(&podSpec)
-		// Apply storageClassDeviceSet PreparePlacement
-		// If nodeAffinity is specified both in the device set and "all" placement,
-		// they will be merged.
+		// when "onlyApplySpecificOSDPlacement" false, in case of NodeAffinity it will merge both spec.placement.All() and storageClassDeviceSets.preparePlacement,
+		// in case of other preparePlacement rule like PodAffinity, PodAntiAffinity... it will override last prepareplacement with the current preparePlacement applied,
+		// technically in another word for prepareplacement apart from NodeAffinity, it will same as "onlyApplySpecificOSDPlacement" true as storageClassDeviceSets.preparePlacement
+		// will override spec.placement.All(). See ApplyToPodSpec().
+
+		// when "onlyApplySpecificOSDPlacement" true, it will only apply storageClassDeviceSets.preparePlacement
+		if !c.spec.Storage.OnlyApplySpecificOSDPlacement {
+			// apply spec.placement.All()
+			c.spec.Placement.All().ApplyToPodSpec(&podSpec)
+		}
+		// apply storageClassDeviceSets.preparePlacement
 		osdProps.getPreparePlacement().ApplyToPodSpec(&podSpec)
 	} else {
-		p := cephv1.GetOSDPlacement(c.spec.Placement)
-		p.ApplyToPodSpec(&podSpec)
+		// when "onlyApplySpecificOSDPlacement" false, in case of NodeAffinity it will merge both spec.placement.All() and spec.Placement.prepareosd,
+		// in case of other preparePlacement rule like PodAffinity, PodAntiAffinity... it will override last preparePlacement with the current preparePlacement applied,
+		// technically in another word for preparePlacement apart from NodeAffinity, it will same as "onlyApplySpecificOSDPlacement" true as spec.Placement.prepareosd will
+		// override spec.placement.All(). See ApplyToPodSpec().
+
+		// when "onlyApplySpecificOSDPlacement" true, it will only apply spec.placement.prepareosd
+		if !c.spec.Storage.OnlyApplySpecificOSDPlacement {
+			// apply spec.placement.All()
+			c.spec.Placement.All().ApplyToPodSpec(&podSpec)
+		}
+		// apply spec.placement.prepareosd
+		c.spec.Placement[cephv1.KeyOSDPrepare].ApplyToPodSpec(&podSpec)
```

Comment:
@sp98 @travisn can you give it a try to see if it is clear... I tried to make it as simple as possible.

And, also I have not tested my changes, will do once the approach is verified.
The reason for not testing is I'm not using `GetOSDPlacement()` as it will make it more complex and doing similar as pvc case.

Reply:
What about the following explanation?
```
The placement for OSDs is computed from several different places: 
- For non-PVCs: `placement.all` and `placement.osd`
- For PVCs: `placement.all` and inside the storageClassDeviceSet from the `placement` or `preparePlacement`

The placement from these sources will be merged by default (if onlyApplySpecificOSDPlacement is false).
If the `placement.all` should not be merged to the OSD placement, onlyApplySpecificOSDPlacement should be set to true. 
```